content/en/blog/test-coverage/index.md

```diff
@@ -0,0 +1,141 @@
+export GOCOVERDIR=”./coverage”
+
+# run binary (and the application tests)
+./program
+
+# convert and analyze result files
+go tool covdata textfmt –i=$GOCOVERDIR –o coverage.txt
+```
+
+Using the new feature means we can include the test coverage of the application test runs in the automated monitoring scope. We put this new feature to use in the Findy Agency project.
+
+## Instrumenting the Container Binary
+
+Measuring the test coverage for application tests meant we needed to refactor our application testing
+pipeline so that CI could gather the data and send it to the analysis tool.
+
+{{< imgproc steps Fit "925x925" >}}
+<em>Acceptance application test set for Findy Agency microservice are run using
+<a href="https://github.com/findy-network/e2e-test-action" target="_blank">a dedicated GitHub Action</a>.
+The action is capable of building and running the container with instrumented binary and reporting the
+coverage data back to the parent job.</em>
+{{< /imgproc >}}
+
+* We use [a shared GitHub action](https://github.com/findy-network/e2e-test-action) to execute
+the same application acceptance test set for all
+our microservices. The action handles setting up the agency environment and running the test set.
+* Service Docker images used in the application tests must have the instrumented binary.
+We modified [the Docker image](https://github.com/findy-network/findy-agent-vault/blob/master/Dockerfile)
+build phase to include the new `-cover` flag.
+* We define the `GO_COVERDIR` environment variable in the service image,
+and the environment definition maps that container path to a folder in the local environment.
+* Once the agency is running, GitHub Action executes the tests. After test execution,
+[CI copies the result files](https://github.com/findy-network/findy-agent-vault/blob/cda27be1aaacdf1e73f8c879408134ce1f2fa076/.github/workflows/test.yml#L30),
+converts and sends them to the analysis tool as with unit tests.
+
+{{< imgproc dockerfile Fit "625x625" >}}
+<em>Changes required to service Dockerfile.</em>
+{{< /imgproc >}}
+
+## Increased Coverage
+
+The change has the expected effect: coverage increased in our repositories where we introduced
+this change. The coverage counts now in previously unmonitored code blocks, such as code in the `main`
+function that was missing unit tests. The change improves our understanding of our testing scope and
+helps us to keep our automated test set in better shape.
+
+You can inspect the approach, for example, in the [findy-agent-vault](https://github.com/findy-network/findy-agent-vault/tree/master#findy-agent-vault)
+repository CI configuration.
+
+{{< youtube EwCFRVkqHic >}}
+*Watch my "Boosting Test Coverage for Microservices" talk on GopherCon UK on YouTube.*
+
+**Note!** The coverage data is lost, if the process does not exit "normally". Therefore, the service needs to
+have graceful shutdown in place, and the Docker container needs to pass the `SIGTERM` signal to
```

Comment:
.. the Docker container needs to pass.. -> maybe: container runtime needs to pass.. or 

It would be cool if the big topic of the graceful shutdown were when it's done, e.g., gRPC service/server. It has to wait for those who are still in progress but do not accept new ones, etc. That could be an exciting blog post. Maybe someone has written something so good that it would be worth referencing.